include/graphblas/algorithms/gmres.hpp

```diff
@@ -0,0 +1,428 @@
+
+/*
+ *   Copyright 2022 Huawei Technologies Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * @author Denis Jelovina
+ */
+
+#ifndef _H_GRB_ALGORITHMS_GMRES
+#define _H_GRB_ALGORITHMS_GMRES
+
+#include <cstdio>
+#include <complex>
+
+#include <graphblas.hpp>
+#include <graphblas/utils/iscomplex.hpp>
+
+namespace grb {
+
+	namespace algorithms {
+
+		/**
+		 * Solves a linear system \f$ b = Ax \f$ with \f$ x \f$ unknown
+		 * by the gmres method on general fields.
+		 *
+		 * Preconditioning is possible by using matrix M.
+		 *
+		 * @tparam descr        The user descriptor
+		 * @tparam IOType       The input/output vector nonzero type
+		 * @tparam ResidualType The type of the residual
+		 * @tparam NonzeroType  The matrix nonzero type
+		 * @tparam InputType    The right-hand side vector nonzero type
+		 * @tparam Ring         The semiring under which to perform GGMRES
+		 * @tparam Minus        The minus operator corresponding to the inverse of the
+		 *                      additive operator of the given \a Ring.
+		 * @tparam Divide       The division operator corresponding to the inverse of
+		 *                      the multiplicative operator of the given \a Ring.
+		 *
+		 * Valid descriptors to this algorithm are:
+		 *   -# descriptors::no_casting
+		 *   -# descriptors::transpose
+		 *
+		 * By default, i.e., if none of \a ring, \a minus, or \a divide (nor their
+		 * types) are explicitly provided by the user, the natural field on double
+		 * data types will be assumed.
+		 *
+		 * \note An abstraction of a field that encapsulates \a Ring, \a Minus, and
+		 *       \a Divide may be more appropriate. This will also naturally ensure
+		 *       that demands on domain types are met.
+		 *
+		 *
+		 * @param[in,out] x              On input: an initial guess to the solution.
+		 *                               On output: the last computed approximation.
```

Comment:
In this case it's not though right? (At the end of the algorithm, x must be computed from Q and H?)

Reply:
Correct, in this version `x` is input only. The most common implementations of GMRES include x-update, but we leave this for after ALP-merge.